### PR TITLE
Modified Gradient Clipping to work with real remote/local model parameters

### PR DESCRIPTION
**Pull request by DanyEle**
Fixed the clip_grad_remote method to work with real-life model parameters, so as to work with the SMS prediction dataset and other real RNN/LSTM/GRU models.

### CODE REVIEW

File: test/torch/test_hook.py [hunk 1]
```diff
@@ -293,23 +293,64 @@ def initHidden(self):
         param.data.add_(-learning_rate, param.grad.data)
 
 
-def test_remote_gradient_clipping(workers):
+# Real test case of gradient clipping
```
Comment (mari-linhares) [inline test/torch/test_hook.py]: Move this comment to docstring

Comment (DanyEle) [reply test/torch/test_hook.py]: Done!

File: syft/frameworks/torch/hook/hook.py [hunk 2]
```diff
@@ -1172,23 +1168,30 @@ def clip_grad_norm_remote_(parameters, max_norm, norm_type=2):
                 total_norm = max(p.grad.data.abs().max() for p in parameters)
             else:
                 # Remote PySyft tensor
-                if hasattr(parameters, "child") and isinstance(
-                    parameters.child, syft.frameworks.torch.pointers.pointer_tensor.PointerTensor
+                if hasattr(parameters[0], "child") and isinstance(
```
[the remark this answers is not in the document]
Comment (DanyEle) [reply syft/frameworks/torch/hook/hook.py]: Created a method to create whether a parameter is remote and refactored the whole clip_grad_norm_remote_ method so as to check whether a list of remote parameters was being passed to it.

**Comment by DanyEle** [comment]
@mari-linhares The `clip_grad_norm_remote_` should be much cleaner now, with the introduced checks for remote parameters.

File: syft/frameworks/torch/hook/hook.py [hunk 3]
```diff
@@ -1169,37 +1169,46 @@ def clip_grad_norm_remote_(parameters, max_norm, norm_type=2):
...
-                    parameters.child, syft.generic.pointers.pointer_tensor.PointerTensor
-                ):
+                # all parameters are remote
+                if all([param_is_pointer_tensor(param) for param in parameters]):
                     total_norm = torch.zeros(1)
-                    # Let's send the total norm over to the remote worker where the remote tensor is
-                    total_norm = total_norm.send(worker)
-                # Local PyTorch tensor
+                    # Let's send the total norm over to the remote where the remote tensor is
+                    total_norm = total_norm.send(parameters[0].location)
```
Comment (robert-wagner) [inline syft/frameworks/torch/hook/hook.py]: why is this only sent to one location

File: test/torch/test_hook.py [hunk 4]
```diff
@@ -293,23 +293,68 @@ def initHidden(self):
...
-    remote_vanishing_tensor = vanishing_tensor_test.send(alice)
-    vanishing_remote_tensor_clipped = torch.nn.utils.clip_grad(remote_vanishing_tensor, 2)
-    # Has the remote gradient indeed increased?
-    greater_tensor_check = (vanishing_remote_tensor_clipped > remote_vanishing_tensor).copy().get()
-    one_tensors = torch.ones([1], dtype=torch.uint8)
-    assert torch.eq(greater_tensor_check, one_tensors)
-
-
-def test_local_gradient_clipping():
-    # Vanishing gradient test
```
Comment (robert-wagner) [inline test/torch/test_hook.py]: is there a reason this was removed? it isn't inherently a bad test

File: test/torch/test_hook.py [hunk 5]
```diff
@@ -293,23 +293,68 @@ def initHidden(self):
...
+    local parameters of an RNN
+    """
     alice = workers["alice"]
-    vanishing_tensor_test = torch.Tensor([-9.8367e23])
-    remote_vanishing_tensor = vanishing_tensor_test.send(alice)
-    vanishing_remote_tensor_clipped = torch.nn.utils.clip_grad(remote_vanishing_tensor, 2)
-    # Has the remote gradient indeed increased?
-    greater_tensor_check = (vanishing_remote_tensor_clipped > remote_vanishing_tensor).copy().get()
-    one_tensors = torch.ones([1], dtype=torch.uint8)
-    assert torch.eq(greater_tensor_check, one_tensors)
```
Comment (robert-wagner) [inline test/torch/test_hook.py]: you could change this to greater_tensor_check.all()

File: syft/frameworks/torch/hook/hook.py (hunk 3 above)
Comment (DanyEle) [reply syft/frameworks/torch/hook/hook.py]: Because the location of all parameters in one model is the same. So, I could have picked the location any random parameter within the model.

File: test/torch/test_hook.py (hunk 4 above)
Comment (DanyEle) [reply test/torch/test_hook.py]: Because in a real scenario, this method is supposed to be applied to a generator of parameters of a model and not to one simple tensor. The new tests checks pretty much the same things as the previous anyway.

File: test/torch/test_hook.py (hunk 5 above)
Comment (DanyEle) [reply test/torch/test_hook.py]: Yeah, that's a good point.

Comment (DanyEle) [reply test/torch/test_hook.py]: Next time I create one such test, I'll use your syntax, thanks! It's much more compact than time. However, this test case was removed.

File: test/torch/test_hook.py (hunk 4 above)
Comment (robert-wagner) [reply test/torch/test_hook.py]: There is a difference between unit and integration tests. The deleted test seems a bit closer to a unit test but it seems you replaced it with an integration test

**Comment by DanyEle** [comment]
@robert-wagner Do you think the two previous "unit" tests should be kept?

**Comment by robert-wagner** [comment]
@DanyEle I do

**Comment by DanyEle** [comment]
> @DanyEle I do

I've put back the test cases related to gradient clipping for local and remote tensors, slightly modifying them to introduce checks over the `tensor.grad` property of tensors instead of the tensors themselves.

File: test/torch/test_hook.py [hunk 6]
```diff
@@ -293,23 +293,125 @@ def initHidden(self):
...
+    # Method operates in place
+    nn.utils.clip_grad_norm_(local_tensor, 2)
+    tensor_comparison_grad = torch.Tensor([[4.5, 4.5], [4.5, 4.5]])
+    # Has the gradient decreased w.r.t. the initial value (i.e. was it clipped)?
+    smaller_tensor_check = local_tensor.grad < tensor_comparison_grad
+
     one_tensors = torch.ones([1], dtype=torch.uint8)
-    assert torch.eq(greater_tensor_check, one_tensors)
+
+    assert torch.eq(one_tensors, smaller_tensor_check.all())
```
Comment (robert-wagner) [inline test/torch/test_hook.py]: You don't need the torch.eq or ones tensor. .all returns a boolean (not a tensor)

Comment (DanyEle) [reply test/torch/test_hook.py]: Well, it looks to me like it actually returns a tensor containing a scalar being 0 for False and 1 for True.

image

So I just changed the check into an equality among integers.

**Comment by DanyEle** [comment]
I guess it should be good now @robert-wagner ?
